utils/s2n_fork_detection.c

```diff
@@ -0,0 +1,369 @@
+#define S2N_FORK_DETECT_IGNORE 0
+#define S2N_FORK_DETECT_DO_NOT_IGNORE 1
+static int ignore_wipeonfork_or_inherit_zero_method_for_testing = S2N_FORK_DETECT_DO_NOT_IGNORE;
+static int ignore_pthread_atfork_method_for_testing = S2N_FORK_DETECT_DO_NOT_IGNORE;
+static int ignore_fork_detection_for_testing = S2N_FORK_DETECT_DO_NOT_IGNORE;
+
+#define S2N_FORK_EVENT 0
+#define S2N_NO_FORK_EVENT 1
+
+#define S2N_FORK_DETECT_NOT_ENABLED 0
+#define S2N_FORK_DETECT_ENABLED 1
+
+struct FGN_STATE {
+    /* The current cached fork generation number for this process */
+    uint64_t current_fork_generation_number;
+
+    /* Semaphore controlling access to the shared sentinel and signaling whether
+     * fork detection is enabled or not. We could use zero_on_fork_addr, but
+     * avoid overloading by using an explicit variable.
+     */
+    int is_fork_detection_enabled;
+
+    /* Sentinel that signals a fork event has occurred */
+    volatile char *zero_on_fork_addr;
+
+    pthread_once_t fork_detection_once;
+    pthread_rwlock_t fork_detection_rw_lock;
+};
+
+/* We only need a single statically initialised state. Note, the state is
+ * inherited by child processes.
+ */
+static struct FGN_STATE fgn_state = {
+    .current_fork_generation_number = 0,
+    .is_fork_detection_enabled = S2N_FORK_DETECT_NOT_ENABLED,
+    .zero_on_fork_addr = NULL,
+    .fork_detection_once = PTHREAD_ONCE_INIT,
+    .fork_detection_rw_lock = PTHREAD_RWLOCK_INITIALIZER,
+};
+
+
+/* Can currently never fail. See initialise_fork_detection_methods() for
+ * motivation.
+ */
+static inline void s2n_initialise_wipeonfork_best_effort(void *addr, long page_size)
+{
+#if defined(USE_MADVISE)
+    /* Ignored on purpose. */
+    madvise(addr, (size_t) page_size, MADV_WIPEONFORK);
+#endif
+}
+
+static inline int s2n_initialise_inherit_zero(void *addr, long page_size)
+{
+#if defined(USE_MINHERIT) && defined(MAP_INHERIT_ZERO)
+    POSIX_ENSURE(minherit(addr, pagesize, MAP_INHERIT_ZERO) == 0, S2N_ERR_FORK_DETECTION_INIT);
+#endif
+
+    return S2N_SUCCESS;
+}
+
+static void s2n_pthread_atfork_on_fork(void)
+{
+    /* This zeroises the first byte of the memory page pointed to by
+     * *zero_on_fork_addr. This is the same byte used as fork event detection
+     * sentinel in s2n_get_fork_generation_number(). The same memory page, and in
+     * turn, the byte, is also the memory zeroised by the MADV_WIPEONFORK fork
+     * detection mechanism.
+     *
+     * Aquire locks to be on the safe side. We want to avoid the checks in
+     * s2n_get_fork_generation_number() getting executed before setting the sentinel
+     * flag. The write lock prevents any other thread from owning any other type
+     * of lock.
+     *
+     * pthread_atfork_on_fork() cannot return errors. Hence, there is no way to
+     * gracefully recover if [un]locking fails.
+     */
+    if (pthread_rwlock_wrlock(&fgn_state.fork_detection_rw_lock) != 0) {
+        abort();
+    }
+    *fgn_state.zero_on_fork_addr = 0;
```

Comment:
Is it worth checking that zero_on_fork_addr isn't still NULL before you de-reference it?

Reply:
Hmmm we don't call `abort()` anywhere else in the codebase, but idk how else you could handle failures in this callback :\ Printing something seems better than silently dying.

We don't expect any failures from this callback in a sane situation, right?